text/0000-add-limits-trait.md

```diff
@@ -0,0 +1,133 @@
+- Feature Name: Limits trait for the rust types
+- Start Date: 2017-12-20
+- RFC PR: 
+- Rust Issue: 
+
+# Summary
+This is an RFC to add a universal trait for the type limits.
+
+# Motivation
+The motivation is quite simple: [make an ability to accept template types with with limits by requiring a trait](https://stackoverflow.com/questions/47904954/rust-finding-the-maximum-allowable-value-for-generic-data-type-t) 
+so that it simplifies and generalizes the code. Another motivation is that we have all that `max_value()` and `min_value()` implemented as
+usual methods of a type implementation, generalizing this to a trait makes the code simplier and avoids duplicating code. Also, looking at
+C++ template of `std::numeric_limits` tells us we must have this thing too because it is easier to use.
+
+# Detailed design
+The design is quite simple: put everything related to the limits what can be generalized into a separate trait in the standard library:
+
+```rust
+trait Limits<T> {
+    fn min_value() -> T;
+    fn max_value() -> T;
+}
```

Comment:
In all of your examples below, you've used constant which is known at compile time.
As the function is like `() -> T`, this means that an associated constant can be used instead as shown below.
Have you considered this design and if so, why did you discard it?

```rust
trait Limits<T> {
    const MIN_VALUE: T;
    const MAX_VALUE: T;
}
```

Reply:
Also.. Why a combined trait instead of one trait for `min` and one for `max`?

Reply:
Nice! Don't know why I forgot we are able to do this in traits! This looks nice, but the question then: why do we have `min_value()` and `max_value()` methods for all primitive types for this?

Reply:
My guess is that these inherent methods precede the stabilization of associated constants, but I could be wrong.

Reply:
@Centril I don't quite understand. Do you mean that before we can have a constant, we need a method? Or simply that the associated constants feature didn't exist when the methods were introduced?

Reply:
The latter.

Reply:
Also, I would prefer `MAX ` and `MIN` so we can reuse std::*::MAX, etc...

Reply:
If there is a use case for runtime-determined (non-`const`) max/min values, it is perhaps best if this RFC blocks on https://github.com/rust-lang/rfcs/pull/2237. Then you can have opt-in const-ness.

Reply:
About combined trait: we may do even better: we may have `Limits` trait which just requires `MinValue` and `MaxValue` traits.

Reply:
About function vs constants. I prefer both the flexibility and simpliness. However, neither constants nor functions have both of them:

1. Constants are constant. They are written once and just used later. This is **simple**, but is not **flexible**: we can't write complex code for calculating it without magic.
2. Functions are not **simple** but **flexible**: you may write both constant value and complex logic there, but as a language entity it is more complex, that just constants. The also reason why functions in a trait are worse is that if you really just want to return a constant you still can't write `const fn` because it is unstable yet.

So, the question is open, I guess. If I am wrong anywhere, correct me please.

Reply:
> My guess is that these inherent methods precede the stabilization of associated constants, but I could be wrong.

`min_value` and `max_value` were always intended to be turned into constants, but they were added pre-1.0, long before associated constants were added! We used to have a `Bounded` trait, but they were rolled back after we simplified the numeric API for conservative reasons before stabilizing the standard library. I think it now lives in the `num` crate.